modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,238 @@
+          State(
+            pending = updatedPending,
+            cap = updatedCap,
+            retryMode = updatedRetryMode,
+            noConfirmationsSinceRetryCount = updatedNoConfirmationsSinceRetryCount,
+            backoffExponent = updatedBackoffExponent
+          )
+        }
+      } yield ()
+
+      def processPending: F[Unit] = stateR.get.flatMap { state =>
+        if (state.retryMode) {
+          val toRetry = state.pending.take(state.cap.toInt)
+          logger.warn(s"[RetryMode] Retrying ${toRetry.size} pending binaries").whenA(toRetry.nonEmpty) >> toRetry.traverse_(tracked =>
+            post(tracked.binary)
+          )
+        } else Applicative[F].unit
+      }
+
+      def clearPending: F[Unit] = stateR.set(State.empty)
+
+      private def post(binary: Hashed[StateChannelSnapshotBinary]): F[Unit] = {
+        val sendRetries = 5
+
+        val retryPolicy: RetryPolicy[F] = RetryPolicies.limitRetries(sendRetries)
+
+        def wasSuccessful: Either[NonEmptyList[StateChannelValidationError], Unit] => F[Boolean] =
+          _.isRight.pure[F]
+
+        def onFailure(binaryHashed: Hashed[StateChannelSnapshotBinary]) =
+          (_: Either[NonEmptyList[StateChannelValidationError], Unit], details: RetryDetails) =>
+            logger.info(s"Retrying sending ${binaryHashed.hash.show} to Global L0 after rejection. Retries so far ${details.retriesSoFar}")
+
+        def onError(binaryHashed: Hashed[StateChannelSnapshotBinary]) = (_: Throwable, details: RetryDetails) =>
+          logger.info(s"Retrying sending ${binaryHashed.hash.show} to Global L0 after error. Retries so far ${details.retriesSoFar}")
+
+        retryingOnFailuresAndAllErrors[Either[NonEmptyList[StateChannelValidationError], Unit]](
+          retryPolicy,
+          wasSuccessful,
+          onFailure(binary),
+          onError(binary)
+        )(
+          globalL0ClusterStorage.getRandomPeer.flatMap { l0Peer =>
+            identifierStorage.get.flatMap { identifier =>
+              stateChannelSnapshotClient
+                .send(identifier, binary.signed)(l0Peer)
+                .onError(e => logger.warn(e)(s"Sending ${binary.hash.show} snapshot to Global L0 peer ${l0Peer.show} failed!"))
+                .flatTap {
+                  case Right(_) =>
+                    logger.info(s"Sent ${binary.hash.show} to Global L0 peer ${l0Peer.show}") >>
+                      stateR.update { state =>
+                        val tracked =
+                          state.pending.zipWithIndex.collectFirst { case found @ (tracked, _) if tracked.binary === binary => found }
+                        tracked match {
+                          case Some((tracked, i)) =>
+                            val updatedTracked = tracked.copy(sendsSoFar = NonNegLong.unsafeFrom(tracked.sendsSoFar + 1))
+                            state.copy(pending = state.pending.updated(i, updatedTracked))
+                          case None => state
+                        }
```

Comment:
I agree, if we need to iterate over whole collection because we used `zipWithIndex` then we still have O(n) and we can use `.map` with `if` inside. However maybe we should use another structure like `Vector` because we want to have a constant append and efficient update but we are always updating at most `cap` elements so `O(cap)` in worst case.

Reply:
>how how often is there going to be something sent that isn't "tracked" ?

We always send tracked elements. Elements are always enqueued before posting.